import sys
from PyQt6 import QtWidgets, uic
from PyQt6 import QtWidgets, QtGui
from src.ai_requests import AI
from src.sql.db_api import DB
import os

class MyApp(QtWidgets.QMainWindow):
    def __init__(self):
        """
        –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –≥–ª–∞–≤–Ω–æ–≥–æ –æ–∫–Ω–∞
        """
        super(MyApp, self).__init__()
        uic.loadUi('main.ui', self)

        self.ai = AI()
        self.db = DB()

        self.ai_mode = self.ai.QA

        self.file_path = ''
        file = self.db.get_newest_file()
        if file:
            self.file_path = file.path

        self.ai_chat_context = []
        self.if_main_text_data_saved = True

        self.refresh_main_text()
        self.load_connections()
        self.load_ui()

    def closeEvent(self, event):
        """
        –°–æ—Ö—Ä–∞–Ω—è–µ—Ç –¥–∞–Ω–Ω—ã–µ –∏–∑ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ —Ñ–∞–π–ª–∞, –µ—Å–ª–∏ —Ñ–∞–π–ª –Ω–µ –±—ã–ª —Å–æ—Ö—Ä–∞–Ω—ë–Ω
        """
        if not self.if_main_text_data_saved:
            unsave_data = self.db.add_unsave_data(self.main_text.toPlainText())
            if unsave_data:
                f = self.db.add_file(path=self.file_path, id_unsave_data=unsave_data.id)
                if not f:
                    f = self.db.get_file_by_path(path=self.file_path)
                    self.db.delete_unsave_data_by_id(id=f.id_unsave_data)
                    self.db.update_file_by_id(id=f.id, id_unsave_data=unsave_data.id)
        else:
            file = self.db.get_file_by_path(path=self.file_path)

            if file.id_unsave_data:
                unsave_data = self.db.get_unsave_data_by_id(id=file.id_unsave_data)
                self.db.delete_unsave_data_by_id(unsave_data.id)
                print("—É–¥–∞–ª–µ–Ω–∏–µ unsave data")
            self.db.delete_file(id=file.id)
            print("—É–¥–∞–ª–µ–Ω–∏–µ —Ñ–∞–π–ª–∞ –∏–∑ –±–¥")
            


    def load_connections(self):
        """
        –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –≤—Å–µ—Ö –∫–Ω–æ–ø–æ–∫ –∫ –∏—Ö —Ö–µ–Ω–¥–ª–µ—Ä–∞–º
        """
        self.open_file.clicked.connect(self.handler_load_file)
        self.ai_send_button.clicked.connect(self.handler_ai_send)
        self.main_text.selectionChanged.connect(self.handler_select_text)
        self.main_text.textChanged.connect(self.handler_text_changed)
        self.save_file.clicked.connect(self.handler_save_file)

    def refresh_main_text(self):
        """
        –û–±–Ω–æ–≤–ª—è–µ—Ç main_text –¥–∞–Ω–Ω—ã–º–∏ –∏–∑ —Ñ–∞–π–ª–∞
        """
        file_path = self.file_path

        if os.path.exists(file_path):
            with open(file_path, 'r', encoding='utf-8') as file:
                text = file.read()
                f = self.db.get_file_by_path(path=file_path)
                if f:
                    unsave_data = self.db.get_unsave_data_by_id(id=f.id_unsave_data)
                    if unsave_data:
                        text = unsave_data.text
                        self.if_main_text_data_saved = False

                self.main_text.setPlainText(text)
                file.close()

    def load_ui(self):
        """
        –ó–∞–≥—Ä—É–∑–∫–∞ –Ω–∞—Å—Ç—Ä–æ–µ–∫ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞
        """
        self.ai_history.setReadOnly(True)
        self.selected_text = ""  

    def ai_analysis(self):
        """
        –ê–Ω–∞–ª–∏–∑–∏—Ä—É–µ—Ç –Ω–µ–π—Ä–æ–Ω–∫–æ–π —Ç–µ–∫—Å—Ç
        """
        out = False
        if self.ai_mode == self.ai.QA:
            if len(self.ai_request.toPlainText()) == 0:
                return False
            if len(self.main_text.toPlainText()) == 0:
                return False
            
            out = self.ai.question_answering(self.ai_request.toPlainText(), self.main_text.toPlainText())
        elif self.ai_mode == self.ai.SUMMARING:
            if len(self.selected_text) == 0:
                return False
            
            try:
                out = self.ai.summaring(self.selected_text, int(self.ai_request.toPlainText()))
            except (ValueError) as e:
                out = self.ai.summaring(self.selected_text)

        return f"[ü§ñ] {out}"
    
    def change_ai_mode(self):
        """
        –ú–µ–Ω—è–µ—Ç —Ä–µ–∂–∏–º, –≤ –∫–æ—Ç–æ—Ä–æ–º —â–∞—Å —Ä–∞–±–æ—Ç–∞–µ—Ç –ò–ò
        (–ø–æ —Ñ–∞–∫—Ç—É –º–µ–Ω—è–µ—Ç –Ω–µ–π—Ä–æ–Ω–∫—É, –∫–æ—Ç–æ—Ä–∞—è —â–∞—Å –±—É–¥–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å)
        """
        request = self.ai_request.toPlainText().strip()
        
        if request.isnumeric():
            self.ai_mode = self.ai.SUMMARING
            return
        if len(request) == 0:
            self.ai_mode = self.ai.SUMMARING
            return
        
        self.ai_mode = self.ai.QA

    def add_to_history(self, *texts):
        """
        –î–æ–±–∞–≤–ª—è–µ—Ç —Ç–µ–∫—Å—Ç –≤ –∏—Å—Ç–æ—Ä–∏—é –¥–∏–∞–ª–æ–≥–∞ —Å –Ω–µ–π—Ä–æ–Ω–∫–æ–π
        """
        for text in texts:
            self.ai_chat_context.append(text)
        self.ai_history.setPlainText('\n'.join(self.ai_chat_context))

    def move_cursor_down(self, t):
        """
        –ü—Ä–æ–∫—Ä—É—á–∏–≤–∞–µ—Ç –¥–æ –Ω–∏–∑–∞ –º–µ–Ω—é—à–∫—É
        """
        cursor = QtGui.QTextCursor(t.document())
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        t.setTextCursor(cursor)

    def handler_text_changed(self):
        self.if_main_text_data_saved = False

    def handler_select_text(self):
        """
        –ù–∞–∑–Ω–∞—á–∞–µ—Ç –Ω–æ–≤–æ–µ –∑–Ω–∞—á–µ–Ω–∏–µ –≤—ã–¥–µ–ª–µ–Ω–Ω–æ–≥–æ —Ç–µ–∫—Å—Ç–∞ –≤ –ø–µ—Ä–µ–º–µ–Ω–Ω—É—é
        """
        self.selected_text = self.main_text.textCursor().selectedText()

    def handler_load_file(self):
        """
        –ó–∞–≥—Ä—É–∂–∞–µ—Ç —Ñ–∞–π–ª –ø–æ –≤—ã–±—Ä–∞–Ω–Ω–æ–º—É –ø—É—Ç–∏
        """
        file_path, _ = QtWidgets.QFileDialog.getOpenFileName(
            self,
            "–í—ã–±–µ—Ä–∏—Ç–µ —Ñ–∞–π–ª",
            "",
            "Text Files (*.txt);;All Files (*)"
        )
        
        self.file_path = file_path
        self.refresh_main_text()

    def handler_save_file(self):
        """
        –°–æ—Ö—Ä–∞–Ω—è–µ—Ç –æ—Ç–∫—Ä—ã—Ç—ã–π —Ñ–∞–π–ª
        """
        file_path = self.file_path

        if file_path:
            with open(file_path, 'w', encoding='utf-8') as file:
                file.write(self.main_text.toPlainText())
                self.if_main_text_data_saved = True
                file.close()

    def handler_ai_send(self):
        """
        –û—Ç–ø—Ä–∞–≤–ª—è–µ—Ç –≤ –Ω–µ–π—Ä–æ–Ω–∫—É —Å–æ–æ–±—â–µ–Ω–∏–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
        """
        self.change_ai_mode()

        print(self.ai_mode)
        out = self.ai_analysis()

        if out:
            request = f"[üë§] {self.ai_request.toPlainText()}"
            self.add_to_history(request, out)
            self.ai_request.setPlainText('')
            self.move_cursor_down(self.ai_history)


if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    window = MyApp()
    window.show()
    sys.exit(app.exec())
